conference_scheduler/parameters.py

```diff
@@ -21,11 +50,45 @@ def _max_one_event_per_slot(shape, X):
         yield sum(X[(event, slot)] for event in range(shape.events)) <= 1
 
 
-def constraints(shape, X):
+def _slots_in_session(slot, session_array):
+    """
+    Return the indices of the slots in the same session as slot
+    """
+    return np.nonzero(session_array[slot])[0]
+
+def _events_with_diff_tag(talk, tag_array):
+    """
+    Return the indices of the events with no tag in common as tag
+    """
+    event_categories = np.nonzero(tag_array[talk])[0]
+    return np.nonzero(sum(tag_array.transpose()[event_categories]) == 0)[0]
+
+def _events_in_session_share_a_tag(session_array, tag_array, X):
+    """
+    Constraint that ensures that if a talk is in a given session then it must
+    share at least one tag with all other talks in that session.
+    """
+    event_indices = range(len(tag_array))
+    session_indices = range(len(session_array))
+    for session in session_indices:
+        slots = _slots_in_session(session, session_array)
+        for slot, event in itertools.product(slots, event_indices):
+            other_events = _events_with_diff_tag(event, tag_array)
+            for other_slot, other_event in itertools.product(slots, other_events):
+                if other_slot != slot and other_event != event:
+                    # If they have different tags they cannot be scheduled
+                    # together
+                    yield X[(event, slot)] + X[(other_event, other_slot)] <= 1
+
+
+def constraints(shape, session_array, tag_array, X):
```

Comment:
Sure happy to reorder although `_events_in_sessions_share_a_tag` doesn't take `shape` as an argument (just `session_array` and `tag_array`): would you still like me to tweak it?

Here's what the `arg` iterates through:

```python
 90     args = ((shape, ), (shape, ), (session_array, tag_array)) 
```

So all constraints will take one of those and `X`.